.github/scripts/cache.py

```diff
@@ -0,0 +1,348 @@
+#!/usr/bin/env python3
+"""
+Strict caching script, that uploads to a hardcoded S3 server. Users should set
+an environment variable 'S3_PASSWORD'.
+
+Usage:
+  cache push (cabal|cargo|build) [--empty-pattern-ok] [--write-cache-found] [--overwrite-ok]
+  cache pull (cabal|cargo|build) [--missing-ok] [--overwrite-ok] [--write-cache-found]
+  cache clean
+  cache -h | --help
+  cache --version
+
+Options:
+  -h --help            Show this screen.
+  --version            Show version.
+  --missing-ok         Do not throw an error if a cache is missing
+  --overwrite-ok       Allow file overwrites on pull, or cache overwrites on push
+  --write-cache-found  Writes '0' or '1' to 'cache_found', depending on whether a cache existed
+  --empty-pattern-ok   Allow patterns to not match any files
+"""
+
+# SPDX-FileCopyrightText: 2022 Google LLC
+#
+# SPDX-License-Identifier: Apache-2.0
+
+import datetime
+import dateutil.tz
+import docopt
+import functools
+import glob
+import hashlib
+import itertools
+import os
+import random
+import string
+import subprocess
+import sys
+
+from typing import List, Sequence
+
+PWD = os.getcwd()
+
+GITHUB_ORG="bittide"
+GITHUB_REPO="bittide-hardware"
+
+CACHE_URL="https://hetzner0.bittide.clash-lang.org"
+CACHE_USER="root"
+CACHE_BUCKET=f"github.org/{GITHUB_ORG}/{GITHUB_REPO}"
+
+CLEAR_AFTER_DAYS=7
+CLEAR_AFTER=f"{CLEAR_AFTER_DAYS}d00h00m00s"
+TOUCH_AFTER=datetime.timedelta(days=1)
+
+CARGO_CACHE_BUST = 2
+CARGO_KEY_PREFIX = f"cargo-{CARGO_CACHE_BUST}-"
+CARGO_KEY_PATTERNS = ("**/Cargo.lock", "**/Cargo.toml")
+CARGO_CACHE_PATTERNS = ("~/.cargo",)
+
+CABAL_CACHE_BUST = 2
+CABAL_KEY_PREFIX = f"cabal-{CABAL_CACHE_BUST}-"
+CABAL_KEY_PATTERNS = ("**/cabal.project", "**/cabal.project.freeze")
+CABAL_CACHE_PATTERNS = ("~/.cabal-nix",)
+
+BUILD_CACHE_BUST = 2
+BUILD_KEY_PREFIX = f"build-products-{BUILD_CACHE_BUST}-"
+BUILD_CACHE_PATTERNS = (
+    f"{PWD}/.ghc.environment.*",
+    f"{PWD}/_build/",
+    f"{PWD}/clash-vexriscv/clash-vexriscv/build_out_dir/",
+    f"{PWD}/dist-newstyle/",
+    f"{PWD}/firmware-support/target/",
+    f"{PWD}/host-tools/target/",
```

Comment:
```suggestion
    f"{PWD}/_build/",
    f"{PWD}/clash-vexriscv/clash-vexriscv/build_out_dir/",
    f"{PWD}/dist-newstyle/",
```

After: https://github.com/bittide/bittide-hardware/pull/427. The `.ghc.environment.*` removal is unrelated: I just don't think we need it.